### PR TITLE
secio: Defer nonce verification.

### LINKED ISSUE #658: Don't wait for the nonce verification at the end of the secio handshake
At the end of the secio handshake, we send a packet to the remote containing their nonce, and wait for the remote's packet, in order to verify that the key exchange process worked.

This is an extra round-trip that can be optimized by assuming that the verification will work and make the handshake `Future` succeed, and close the connection/substream if it turns out that it didn't.

### PR DESCRIPTION
**Pull request by twittner**
Closes #658.

### CODE REVIEW

File: protocols/secio/src/codec/decode.rs [hunk 1]
```diff
@@ -42,21 +43,23 @@ pub struct DecoderMiddleware<S> {
...
 }
 
 impl<S> DecoderMiddleware<S> {
     #[inline]
-    pub fn new(
-        raw_stream: S,
-        cipher: StreamCipher,
-        hmac: Hmac,
-    ) -> DecoderMiddleware<S> {
+    pub fn new(raw_stream: S, cipher: StreamCipher, hmac: Hmac) -> DecoderMiddleware<S> {
```
Comment (tomaka) [inline protocols/secio/src/codec/decode.rs]: Why not put the `nonce` as part of the constructor? I don't see why we would change it later.

File: protocols/secio/src/lib.rs [hunk 2]
```diff
@@ -427,19 +427,15 @@ where
     ///
     /// On success, produces a `SecioMiddleware` that can then be used to encode/decode
     /// communications, plus the public key of the remote, plus the ephemeral public key.
-    pub fn handshake<'a>(
-        socket: S,
-        config: SecioConfig,
-    ) -> Box<Future<Item = (SecioMiddleware<S>, PublicKey, Vec<u8>), Error = SecioError> + Send + 'a>
+    pub fn handshake<'a>(socket: S, config: SecioConfig)
+        -> impl Future<Item = (SecioMiddleware<S>, PublicKey, Vec<u8>), Error = SecioError> + Send + 'a
```
Comment (tomaka) [inline protocols/secio/src/lib.rs]: You can remove the `Send` and the `'a` if you use `impl Future`.

File: protocols/secio/src/handshake.rs [hunk 3]
```diff
@@ -310,7 +310,7 @@ impl HandshakeContext<Ephemeral> {
 pub fn handshake<'a, S: 'a>(
     socket: S,
     config: SecioConfig
-) -> Box<Future<Item = (FullCodec<S>, PublicKey, Vec<u8>), Error = SecioError> + Send + 'a>
+) -> impl Future<Item = (FullCodec<S>, PublicKey, Vec<u8>), Error = SecioError> + Send + 'a
```
Comment (tomaka) [inline protocols/secio/src/handshake.rs]: Same, remove the `Send` and `'a`.

File: protocols/secio/src/codec/decode.rs (hunk 1 above)
Comment (tomaka) [reply protocols/secio/src/codec/decode.rs]: Also some documentation saying that `nonce` is what the stream of data has to start with would be nice.

File: protocols/secio/src/codec/mod.rs [hunk 4]
```diff
@@ -109,12 +109,15 @@ pub fn full_codec<S>(
...
     decoding_hmac: Hmac,
+    remote_nonce: Vec<u8>
 ) -> FullCodec<S>
 where
     S: AsyncRead + AsyncWrite,
 {
     let encoder = EncoderMiddleware::new(socket, cipher_encoding, encoding_hmac);
-    DecoderMiddleware::new(encoder, cipher_decoder, decoding_hmac)
+    let mut decoder = DecoderMiddleware::new(encoder, cipher_decoder, decoding_hmac);
+    decoder.remote_nonce(remote_nonce);
```
Comment (dvdplm) [inline protocols/secio/src/codec/mod.rs]: This piece looks a bit odd, seems like the `remote_nonce` could be part of the `new()` call?

**Pull request merged**